### ISSUE TITLE
12 digits for German mobile phone number are not supported

### ISSUE THREAD

**Issue by dbrammertz** [googlei18n/libphonenumber#710]
Used libphonenumber version 7.0.6

The mobile phone numbers in germany can be up to 12 digits. E.g. +49 174 456789012 is a valid number. See also: http://en.wikipedia.org/wiki/Telephone_numbers_in_Europe

**Comment by dbrammertz** [comment]
Proposed patch:

@@ -5550,11 +5550,11 @@
       <availableFormats>
         
-        <numberFormat pattern="(1\d{2})(\d{7,8})">
+        <numberFormat pattern="(1\d{2})(\d{7,9})">
           <leadingDigits>1[67]</leadingDigits>
           <format>$1 $2</format>
         </numberFormat>
-        <numberFormat pattern="(1\d{3})(\d{7})">
+        <numberFormat pattern="(1\d{3})(\d{8})">
           <leadingDigits>15</leadingDigits>
           <format>$1 $2</format>
         </numberFormat>
@@ -5807,15 +5807,15 @@
              -->
         <nationalNumberPattern>
           1(?:
-            5[0-2579]\d{8}|
-            6[023]\d{7,8}|
+            5[0-2579]\d{8,9}|
+            6[023]\d{8,9}|
             7(?:
               [0-57-9]\d?|
               6\d
-            )\d{7}
+            )\d{7,8}
           )
         </nationalNumberPattern>
-        <possibleNumberPattern>\d{10,11}</possibleNumberPattern>
+        <possibleNumberPattern>\d{10,12}</possibleNumberPattern>
         <exampleNumber>15123456789</exampleNumber>
       </mobile>
       <pager>

**Comment by bmarry** [comment]
Hi

Please see the ITU phone plan for Germany

http://www.itu.int/dms_pub/itu-t/oth/02/02/T02020000510002PDFE.pdf

For the prefix 174 the maximum length is 10 digits not 12 as in your
example.

Please supply more evidence that their are live numbers with this prefix of
length 12, in order for us to evaluate an update.

Thank you for email and suggestion

Brendan

On 2 June 2015 at 17:52, dbrammertz <notifications@github.com> wrote:

> Proposed patch:
>
> @@ -5550,11 +5550,11 @@
>
> 
>
>    -
>    - 1[67] $1 $2
>    -
>    - 15 $1 $2 @@ -5807,15 +5807,15 @@ --> 1(?:
>    - 5[0-2579]\d{8}|
>    - 6[023]\d{7,8}|
>    - 5[0-2579]\d{8,9}|
>    - 6[023]\d{8,9}| 7(?: [0-57-9]\d?| 6\d
>    - )\d{7}
>    - )\d{7,8} )
>    - \d{10,11}
>    - \d{10,12} 15123456789
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/googlei18n/libphonenumber/issues/710#issuecomment-107997497>
> .
>



-- 
Brendan Marry | Google International Product Manager EMEA |
bmarry@google.com | +41 79 763 37 13

This email may be confidential or privileged.  If you received this
communication by mistake, please don't forward it to anyone else, please
erase all copies and attachments, and please let me know that it went to
the wrong person.  Thanks.

**Comment by futuretap** [comment]
Those 12 digit cell phone numbers definitely exist, but not only for the 174 prefix. For instance
`+49 157 706816305` is a valid number. I assume that all German cell phone area codes can now have 9 digits after the area code.

**Comment by bmarry** [comment]
The ITU documentation states max and min number of digits are 11 including
the 1577 prefix that leaves 7 digits after the prefix the number you
supplied has 8. This documentation was updated in May 2015.

When I tried to dial the number in your email it does not connect nor does
it have a ringing tone. Can you please check that the number connects
and/or if there is evidence for the Telco E-Plus Mobilfunk GmbH & Co. KG
that their numbers are longer than the ITU stated lenght

Thank you
Brendan


On 5 June 2015 at 12:06, Ortwin Gentz, FutureTap <notifications@github.com>
wrote:

> Those 12 digit cell phone numbers definitely exist, but not only for the
> 174 prefix. For instance
> +49 157 706816305 is a valid number. I assume that all German cell phone
> area codes can now have 9 digits after the area code.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/googlei18n/libphonenumber/issues/710#issuecomment-109247611>
> .
>



-- 
Brendan Marry | Google International Product Manager EMEA |
bmarry@google.com | +41 79 763 37 13

This email may be confidential or privileged.  If you received this
communication by mistake, please don't forward it to anyone else, please
erase all copies and attachments, and please let me know that it went to
the wrong person.  Thanks.

**Comment by futuretap** [comment]
Brendan, I believe you're right and the number is indeed invalid.

**Comment by dbrammertz** [comment]
Hi Brendan,

thank you for your patience. Finally I got the real phone number from our support. Indeed it is not a regular number but a mailbox number. The network operator add 2 digits between the prefix and the digits to signal that the mailbox is to be called. In my scenario the number was 0174 50 9 xxx xxx, the 50 means mailbox so the real number is 0174 9 xxx xxx.

In the configuration there is a simular case for e-plus:
```
       <voicemail>
        <nationalNumberPattern>17799\d{7,8}</nationalNumberPattern>
        <possibleNumberPattern>\d{12,13}</possibleNumberPattern>
        <exampleNumber>177991234567</exampleNumber>
      </voicemail>
```

But there are more of that:

```
0151 -13 - number
0152 - 0-50 - number (if proted 0152-0-55-number)
0157 - 7-99 - number
0159 - 33 - number
0160 - 13 - number
0162 - 50 - number (if proted 0162-55-number)
0163 - 99 - number
0170 - 13 - number
0171 - 13 - number
0172 - 50 - number (if proted 0172-55-number)
0173 - 50 - number (if proted 0173-55-number)
0174 - 50 - number (if proted 0174-55-number)
0175 - 13 - number
0176 - 33 - number
0177 - 99 - number
0178 - 99 - number
0179 - 33 - number
```

**Comment by roes** [comment]
Hi,
Do you have any supporting evidence that these prefixes are all in use and voicemail? For example an official government or telecom provider document?

**Comment by dbrammertz** [comment]
Finally I found an official paper:

http://www.bundesnetzagentur.de/SharedDocs/Downloads/EN/BNetzA/Areas/Telecommunications/NumberManagement/MobileServices/Mobileservices_Numbering%20Plan200913.pdf?__blob=publicationFile&v=1

In chapter "2.5 Numbers for voicemail services" you can find "A voicemail number is formed by inserting two digits in the subscriber number (infix)."

**Comment by mihaelacr-google** [comment]
Hi!

In the documentation you provided there is no evidence of associating the infix '50' with the 171, 172, 173 and 174 codes, but rather 55. Everything else looks good.

**Comment by dbrammertz** [comment]
You are right in the official document I can't find an evidence about that infix for the prefixes 152, 162, 172, 173 and 174 (never stated 171). All prefixes belong to vodaphone. I checked their help system for the infixes and found: http://www.vodafone.de/infofaxe/317.pdf, "Nachrichten direkt auf die Vodafone Mailbox bzw. ProfiMailbox sprechen"

It seems that the "50" is not a mailbox but a service to leave messages on the mailbox of somebody without ringing his phone.

Not quite sure wether it fits in this library although it would be quite handy for us, since we are validating if the given phone number are valid targets.

**Comment by icywolfy** [comment]
The original ticket concerning Voicemail Access Numbers is at : https://github.com/googlei18n/libphonenumber/issues/439
That has a bit more of a summary. 
Only difference is the "fax" infix (+50 on the +55 infix vorwahls) is used to direct-dial mailboxes for those vorwahls.

(Other issues dealing with German Voicemail Infixes in libphonenumber:
https://github.com/googlei18n/libphonenumber/issues/376#issuecomment-65484052
https://github.com/googlei18n/libphonenumber/issues/439
https://github.com/googlei18n/libphonenumber/issues/359#issuecomment-65483683
)

**Comment by padmaksha** [comment]
Hi ,

We are working on resolving  this issue and have come across one constraint.
There is a difference in the position the infix is inserted for three prefixes 152, 157 and 159 between the official document [bundesnetzagentur](http://www.bundesnetzagentur.de/SharedDocs/Downloads/EN/BNetzA/Areas/Telecommunications/NumberManagement/MobileServices/Mobileservices_Numbering%20Plan200913.pdf?__blob=publicationFile&v=1) and [t-mobile's website](https://www.t-mobile.de/business/faq/1,16589,14441-_,00.html?c=102).

The official doc [bundesnetzagentur](http://www.bundesnetzagentur.de/SharedDocs/Downloads/EN/BNetzA/Areas/Telecommunications/NumberManagement/MobileServices/Mobileservices_Numbering%20Plan200913.pdf?__blob=publicationFile&v=1) defines the basic structure of mobile numbers as consisting of a two digit service code(15,16 or 17) , a two or three digit block code and a terminal equipment number(SN). For formation of voicemail numbers infix is inserted at appropriate position.

The official doc [bundesnetzagentur](http://www.bundesnetzagentur.de/SharedDocs/Downloads/EN/BNetzA/Areas/Telecommunications/NumberManagement/MobileServices/Mobileservices_Numbering%20Plan200913.pdf?__blob=publicationFile&v=1) states that for these three prefixes 152, 157 and 159  the infix is inserted between the block code and the terminal equipment number for two digit block code and  between the second and third digits of the block code  for three digit block code. This relates to formation of voicemail numbers as shown below.

| Sl. No | Prefix   | ServiceCode | BlockCode | Infix | Format    | Length of Block Code |
|--------|----------|-------------|-----------|-------|-----------|----------------------|
| 1      | 152\d    | 15          | 2\d       | 55    | 152\d55   | 2 digit Block code   |
| 2      | 157\d    | 15          | 7\d       | 99    | 157\d99   |                      |
| 3      | 159\d    | 15          | 9\d       | 33    | 159\d33   |                      |
| 4      | 152\d{2} | 15          | 2\d{2}    | 55    | 152\d55\d | 3 Digit Block Code   |
| 5      | 157\d{2} | 15          | 7\d{2}    | 99    | 157\d99\d |                      |
| 6      | 159\d{2} | 15          | 9\d{2}    | 33    | 159\d33\d |                      |

The  [t-mobile's website](https://www.t-mobile.de/business/faq/1,16589,14441-_,00.html?c=102) on the other hand states that the prefix is always three digits and  that we should insert the infix after the prefixes where prefix is defined as 152, 157 and 159. This relates to formation of voicemail numbers as below .

| Sl. No | Prefix | Infix | Format |
|--------|--------|-------|--------|
| 1      | 152    | 55    | 15255  |
| 2      | 157    | 99    | 15799  |
| 3      | 159    | 33    | 15933  |

As it is evident from the two tables above the two documents contradict each other and there is a discrepancy in the position of the inserted infix as well as the format of the voicemail number formed.

It would be of great help if anyone can provide any documentation or evidence that clears
this contradiction so we can implement the change at the earliest. All efforts will be greatly appreciated.

Thanks and Regards

**Comment by cfkoch** [comment]
It's always 0152x-infix;  
Documentation almost universally just lists the "if your phone number starts with : 0152" use "55",   they (rightfully), don't special case the 01520, 01521, 01522, etc when the infix is the same for the entire subset of 0152.  It's just known that you put the infix after the mobile prefix. 

(Hence the formatting issue in another ticket, because even though Vodaphone originally sold "0152" as the prefix, and a mobile number of 0123456 (ie: 0152-0123456); but the voicemail access is 0152 0-55-123456 -- that failed miserably and number formatting/mindset is   01520-123456, with the infix in the normal position after the (then) exceptional 4-digit vorwahl)


It's common knowledge, but doing a quick search:
- https://www.simyo.de/faq/mailbox-abhren-4010.html
- http://hilfe-center.1und1.de/mobilfunk-c83940/mobil-telefonieren-c84077/1und1-mailbox-d-netz-c85253/1und1-mailbox-abhoeren-d-netz-a787139.html
- https://hilfe.o2online.de/docs/DOC-1067
- http://www.areamobile.de/community/netzbetreiber-provider/121441-welches-netz-hinter-welcher-nummer.html
- http://webcache.googleusercontent.com/search?q=cache:a77_o6g-0sMJ:mobilfaq.in-ulm.de/umlbox.faq.html+&cd=1&hl=en&ct=clnk&gl=us
- https://www.congstar-forum.de/mobilfunk/archiv-mobilfunk/39755-mailbox-aus-anderen-netzen-anrufen-bei-portierten-rufnummern/
...

**Comment by icywolfy** [comment]
It's always 0152x-infix;

Documentation almost universally just lists the "if your phone number starts with : 0152" use "55", they (rightfully), don't special case the 01520, 01521, 01522, etc when the infix is the same for the entire subset of 0152. It's just known that you put the infix after the mobile prefix.

(Hence the formatting issue in another ticket, because even though Vodaphone originally sold "0152" as the prefix, and a mobile number of 0123456 (ie: 0152-0123456); but the voicemail access is 0152 0-55-123456 -- that failed miserably and number formatting/mindset is 01520-123456, with the infix in the normal position after the (then) exceptional 4-digit vorwahl)

It's common knowledge, but doing a quick search:
- https://www.simyo.de/faq/mailbox-abhren-4010.html
- http://hilfe-center.1und1.de/mobilfunk-c83940/mobil-telefonieren-c84077/1und1-mailbox-d-netz-c85253/1und1-mailbox-abhoeren-d-netz-a787139.html
- https://hilfe.o2online.de/docs/DOC-1067
- http://www.areamobile.de/community/netzbetreiber-provider/121441-welches-netz-hinter-welcher-nummer.html
- http://webcache.googleusercontent.com/search?q=cache:a77_o6g-0sMJ:mobilfaq.in-ulm.de/umlbox.faq.html+&cd=1&hl=en&ct=clnk&gl=us
- https://www.congstar-forum.de/mobilfunk/archiv-mobilfunk/39755-mailbox-aus-anderen-netzen-anrufen-bei-portierten-rufnummern/ 
- ...

**Comment by dbrammertz** [comment]
Hi Padmaksha,

both documents states the same  the one from the Bundesnetzagentur gives more information and adds more complexity. Basically the inset is inserted between the block code and terminal equipment number. For 15x the block may have the size 2 to 3.
To clarify the situation the chapter "2.5 Numbers for voicemail services" states which Service code and 
block code is used for which infix. 

You can see that this table in chapter 2.5 is equal to the T-Mobile listing.

**Comment by missinglink** [comment]
I can confirm that I have a German mobile SIM, my number is: 01766707xxxx

**Comment by padmaksha** [comment]
Hi,

Thanks for the detailed explanations.
We're working on it. We'll update this thread when it's released.

**Comment by penmetsaa** [comment]
Fixed in release 7.3.1:
https://libphonenumber.appspot.com/phonenumberparser?number=%2B49+152055+1230000
